File: src/commands/avatar.rs
/// /avatar [user] - display a user's avatar in full resolution.
use crate::components::emoji::{header, hint, stat, Colour, E};
use crate::components::v2::{respond_to_interaction, ButtonStyle, FadeResponse};
use crate::error::{BotError, BotResult};
use crate::state::AppState;
use serenity::{
    model::application::{ComponentInteraction, ResolvedOption, ResolvedValue},
    prelude::*,
};
use std::sync::Arc;
use tokio::sync::RwLock;

pub async fn run(
    ctx: &Context,
    cmd: &crate::commands::context::CommandContext<'_>,
    _state: Arc<RwLock<AppState>>,
    _args: &[&str],
) -> BotResult<()> {
    let mut guild_avatar = None;
    let target = match cmd {
        crate::commands::context::CommandContext::Slash(c) => c.data.options().iter().find_map(|opt| {
            if opt.name == "user" {
                if let ResolvedOption { value: ResolvedValue::User(u, _), .. } = opt {
                    return Some((*u).clone());
                }
            }
            None
        }).unwrap_or_else(|| {
            guild_avatar = c.member.as_ref().and_then(|m| m.avatar_url()).map(|u| u.replace("size=1024", "size=4096"));
            c.user.clone()
        }),
        crate::commands::context::CommandContext::Prefix(msg) => {
            if let Some(user) = msg.mentions.first() {
                user.clone()
            } else {
                msg.author.clone()
            }
        }
    };

    if guild_avatar.is_none() && cmd.guild_id().is_some() {
        if let Ok(m) = ctx.http.get_member(cmd.guild_id().unwrap(), target.id).await {
            guild_avatar = m.avatar_url().map(|url| url.replace("size=1024", "size=4096"));
        }
    }

    let display_name = target.global_name.as_deref().unwrap_or(&target.name).to_string();

    let base_url = target.avatar_url().unwrap_or_else(|| target.default_avatar_url());
    let png_hd   = base_url.replace("size=1024", "size=4096").replace(".webp", ".png");
    let has_anim = target.avatar.as_ref().map(|a| a.to_string().starts_with("a_")).unwrap_or(false);
    let gif_url  = has_anim.then(|| {
        base_url.replace("size=1024", "size=4096").replace(".webp", ".gif")
    });

    let main_url = gif_url.unwrap_or(png_hd.clone());

    let response = FadeResponse::new()
        .container(None, |c| {
            let c = c.text(header(E::BRAND, &display_name))
                     .text(stat(E::ID, "User ID", target.id))
            .separator(true);

            let c = c.media_gallery(|g| {
                g.item(&main_url, None)
            })
            .separator(true);

            c.action_row(|r| {
                let r = r.link(&main_url, "View Avatar");
                if guild_avatar.is_some() {
                    let custom_id = format!("avatar_SERVER_{}", target.id);
                    r.button(&custom_id, "Server Avatar", ButtonStyle::Primary)
                } else {
                    r
                }
            })
        });

    cmd.respond(ctx, &response).await?;
    Ok(())
}

File: ts/src/commands/avatar.ts
import { Declare, Command, type CommandContext, Options, createStringOption, createUserOption } from 'seyfert';
import { FadeResponse, ButtonStyle } from '../components/v2';
import { header, stat, E } from '../components/emoji';

const options = {
  user: createUserOption({
    description: 'The user to get the avatar of',
    required: false,
  }),
};

@Declare({
  name: 'avatar',
  description: 'Display a user\'s avatar in full resolution.',
})
@Options(options)
export default class AvatarCommand extends Command {
  override async run(ctx: CommandContext<typeof options>) {
    const targetUser = ctx.options.user ?? ctx.author;
    let guildAvatar: string | undefined;

    if (ctx.guildId) {
      try {
        const member = ctx.options.user 
            ? await ctx.client.members.fetch(ctx.guildId, targetUser.id)
            : ctx.member;
        
        if (member) {
          const url = member.avatarURL();
          if (url) {
            guildAvatar = url.replace('size=1024', 'size=4096');
          }
        }
      } catch (err) {
        // Ignored
      }
    }

    const displayName = targetUser.globalName || targetUser.username;
    const baseUrl = targetUser.avatarURL() || targetUser.defaultAvatarURL();
    
    // Attempt to convert webp to png/gif
    let pngHd = baseUrl.replace('size=1024', 'size=4096').replace('.webp', '.png');
    // If avatar starts with a_ it is animated. Seyfert handles it in avatarURL(true)? 
    // Usually discord urls end with .gif if animated
    let mainUrl = baseUrl;
    if (baseUrl.includes('.webp')) {
      if (targetUser.avatar?.startsWith('a_')) {
        mainUrl = baseUrl.replace('size=1024', 'size=4096').replace('.webp', '.gif');
      } else {
        mainUrl = pngHd;
      }
    } else {
      mainUrl = baseUrl.replace('size=1024', 'size=4096');
    }

    const response = new FadeResponse()
      .container(undefined, c => {
        c.text(header(E.BRAND, displayName))
         .text(stat(E.ID, 'User ID', targetUser.id))
         .separator(true)
         .mediaGallery(g => g.item(mainUrl, undefined))
         .separator(true);

        c.actionRow(r => {
          r.link(mainUrl, 'View Avatar');
          if (guildAvatar) {
            r.button(`avatar_SERVER_${targetUser.id}`, 'Server Avatar', ButtonStyle.Primary);
          }
          return r;
        });
        return c;
      });

    await ctx.write(response.toMessage());
  }
}
